src/HexagonOptimize.cpp

```diff
@@ -1760,38 +1785,84 @@ class VtmpyGenerator : public IRMutator {
-
-                            vtmpy_exprs.emplace_back(native_interleave(Call::make(op->type,
-                                "halide.hexagon.vtmpy" + vtmpy_suffix,
-                                { mpys[v0_idx].first, mpys[v2_idx].first,
-                                  mpys[v0_idx].second, mpys[v1_idx].second },
-                                Call::PureExtern)));
-                            // As we cannot test the same indices again
-                            i = i+2;
+                            if (vrmpy_suffix.empty()) {
+                                if (is_const(mpys[v2_idx].second, 1)) {
+                                    // vtmpy opporrunity
+                                    sw_indices[v0_idx] = true;
+                                    sw_indices[v1_idx] = true;
+                                    sw_indices[v2_idx] = true;
+
+                                    sw_exprs.emplace_back(native_interleave(Call::make(op->type,
+                                        "halide.hexagon.vtmpy" + vtmpy_suffix,
+                                        { mpys[v0_idx].first, mpys[v2_idx].first,
+                                          mpys[v0_idx].second, mpys[v1_idx].second },
+                                        Call::PureExtern)));
+                                    i = i+2;
+                                }
+                            } else {
+                                // vrmpy opporrunity
+
+                                // Vrmpy helps in sliding window ops of the form:
+                                // b0*v0 + b1*v1 + b2*v2 + b3*v3. In case we cannot
+                                // find v3, we set b3=0.
+                                Expr b3 = make_zero(mpys[v0_idx].second.type());
+                                if (i+3 < vec_size) {
+                                    Expr v3 = iter->second[i+3].first;
+                                    size_t v3_idx = iter->second[i+3].second;
+                                    if (is_base_shifted(v3, v2, 1)) {
+                                        // Try finding the 4th vector. If not, use b3 = 0.
+                                        sw_indices[v3_idx] = true;
+                                        b3 = mpys[v3_idx].second;
+                                    }
+                                }
+
+                                Expr shifted_expr = shift_load_by_2(mpys[v2_idx].first);
+
+                                sw_indices[v0_idx] = true;
+                                sw_indices[v1_idx] = true;
+                                sw_indices[v2_idx] = true;
+
+                                Expr b0123 = Shuffle::make_interleave({ mpys[v0_idx].second,
+                                                mpys[v1_idx].second, mpys[v2_idx].second, b3 });
+                                b0123 = simplify(b0123);
+                                b0123 = reinterpret(Type(b0123.type().code(), 32, 1), b0123);
```

Comment:
Might be better to just use Int(32) for clarity and to constrain the code. This has to be 4 x Int(8) right?